pkg/policy/k8s/cilium_network_policy.go

```diff
@@ -162,26 +161,22 @@ func (p *policyWatcher) upsertCiliumNetworkPolicyV2(cnp *types.SlimCNP, initialR
 
 	scopedLog.Debug("Adding CiliumNetworkPolicy")
 
-	rules, policyImportErr := cnp.Parse()
-	if policyImportErr == nil {
-		_, policyImportErr = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
-			Source:              source.CustomResource,
-			ProcessingStartTime: initialRecvTime,
-			Resource:            resourceID,
-			ReplaceByResource:   true,
-		})
-	}
-
-	if policyImportErr != nil {
-		scopedLog.WithError(policyImportErr).Warn("Unable to add CiliumNetworkPolicy")
-	} else {
-		scopedLog.Info("Imported CiliumNetworkPolicy")
+	rules, err := cnp.Parse()
+	if err != nil {
+		scopedLog.WithError(err).Warn("Unable to add CiliumNetworkPolicy")
+		return fmt.Errorf("failed to parse CiliumNetworkPolicy %s/%s: %w", cnp.ObjectMeta.Namespace, cnp.ObjectMeta.Name, err)
 	}
-
-	return policyImportErr
+	p.policyImporter.UpdatePolicy(&policytypes.PolicyUpdate{
```

Comment:
So the previous `PolicyAdd` was blocking. Now we're just enqueuing it, without waiting for the policy insertion to finish. During regular operations that's fine, but during startup I think that might be an issue.

Since we declare the K8s caches as ready as soon as we've called `UpdatePolicy` for every observed K8s NetPol, I believe there is a chance that endpoint regeneration starts before the enqueued policies have been ingested. That would then lead to drops, since some K8s policies might not be part of the post-restart BPF policy map.

How are we ensuring all policies are actually in the policy repository before we do the initial endpoint regen?

Reply:
Ohh, very good point!

Reply:
@gandro I added a slightly-complicated bit to track in-flight sync requests. It works, though I don't 100% love it.

One cute bit to cut down on boilerplate I considered was treating "revision 0 on the channel as increment, actually", since the policy repository always starts with revision number 1. If you find the number of variables distasteful, I can do this at the cost of a magic number.

Reply:
Thanks! What you came up with seems pragmatic and correct!

Eventually we should re-consider the whole K8s caches infrastructure (it's soo error prone) and introduce a separation of concerns (i.e. have a generic "policies have been ingested" event). That way we might not have to track three different counters. But that's something we should tackle separately.